performance/multi-node/rdm_bw.c

```diff
@@ -0,0 +1,485 @@
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AWV
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <unistd.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <math.h>
+#include <assert.h>
+#include <string.h>
+
+#include <rdma/fabric.h>
+#include <rdma/fi_domain.h>
+#include <rdma/fi_errno.h>
+#include <rdma/fi_endpoint.h>
+#include <rdma/fi_cm.h>
+#include <rdma/fi_rma.h>
+
+#include "ct_utils.h"
+
+#define MAX_ALIGNMENT 65536
+#define MAX_MSG_SIZE (1<<22)
+#define MYBUFSIZE (MAX_MSG_SIZE + MAX_ALIGNMENT)
+
+#define TEST_DESC "Libfabric Bandwidth/Latency Test"
+#define HEADER "# " TEST_DESC " \n"
+#ifndef FIELD_WIDTH
+#   define FIELD_WIDTH 20
+#endif
+#ifndef FLOAT_PRECISION
+#   define FLOAT_PRECISION 2
+#endif
+
+int loop = 100000;
+int window_size = 1;
+int skip = 1000;
+
+int loop_large = 20;
+int window_size_large = 64;
+int skip_large = 2;
+
+int large_message_size = 8192;
+
+char s_buf_original[MYBUFSIZE];
+char r_buf_original[MYBUFSIZE];
+
+static int rx_depth = 512;
+
+struct fi_info *fi, *hints;
+struct fid_fabric *fab;
+struct fid_domain *dom;
+struct fid_ep *ep;
+struct fid_av *av;
+struct fid_cq *rcq, *scq;
+struct fid_mr *r_mr, *l_mr;
+struct fi_context fi_ctx_send;
+struct fi_context fi_ctx_recv;
+struct fi_context fi_ctx_av;
+
+void *addrs;
+fi_addr_t *fi_addrs;
+
+typedef struct buf_desc {
+	uint64_t addr;
+	uint64_t key;
+} buf_desc_t;
+
+buf_desc_t *rbuf_descs;
+
+int myid, numprocs;
+
+static void cq_readerr(struct fid_cq *cq, const char *cq_str)
+{
+	struct fi_cq_err_entry cq_err;
+	const char *err_str;
+	int ret;
+
+	ret = fi_cq_readerr(cq, &cq_err, 0);
+	if (ret < 0) {
+		ct_print_fi_error("fi_cq_readerr", ret);
+	} else {
+		err_str = fi_cq_strerror(cq, cq_err.prov_errno, cq_err.err_data,
+					NULL, 0);
+		fprintf(stderr, "%s: %d %s\n", cq_str, cq_err.err,
+			fi_strerror(cq_err.err));
+		fprintf(stderr, "%s: prov_err: %s (%d)\n", cq_str, err_str,
+			cq_err.prov_errno);
+	}
+}
+
+/*
+ * fi_cq_err_entry can be cast to any CQ entry format.
+ */
+static int ct_wait_for_comp_omb(struct fid_cq *cq, int num_completions)
```

Comment:
There's a bit of history here, but the end result was that this function was copied in all the source files in this directory.  I think I'll just inline the code in the benchmarks.